Tar: Write unused GNU tar entry size, ctime and atime bytes in the expected format

carlossanlop opened the pull request:
This PR fixes two bugs in the GNU format:

1. The atime and ctime fields (unique to this format) need to be written as null characters if their value is the Unix Epoch. The bug was that we were writing them as "0" characters as other GNU metadata fields. This was causing tools like 7-zip to show an extra unexpected folder with a name formed by just "0" characters. Other tools were unable to read these archives as they were considered corrupt.

    Added sync and async tests to validate directly in the memory stream that when atime and ctime represent the Unix Epoch, they are indeed written as null characters.

2. The unused bytes of the size metadata field should be filled with "0" characters. The bug was that we were filling them with nulls like in the other formats.
    Added sync and async tests to validate directly in the memory stream that the unused bytes in the size field are indeed written as "0" characters.

    Added sync and async tests to validate directly in the memory stream that when atime and ctime represent the Unix Epoch, they are indeed written as null characters.

Manual test: reading entries in a GNU archive generated by another tool, then rewriting each entry into a new archive created with TarWriter. Then opening it with 7-zip.

Before - Note how 7-zip did its best to deduce the minimum required fields, but most was empty:
![image](https://github.com/user-attachments/assets/9a6ef1b0-8d43-4052-a9c8-61ea301c9978)

After - Note that 7-zip now shows many other fields that it was unable to read before:
![image](https://github.com/user-attachments/assets/372cda56-3c9a-492b-a4f9-1e598bee3e90)

carlossanlop committed:
In the GNU tar entry format, the unused bytes of the size metadata field should be filled with "0" characters. The bug was that we were filling them with nulls like in the other formats.
Added sync and async tests to validate directly in the memory stream that the unused bytes in the size field are indeed written as "0" characters.

carlossanlop committed:
In the GNU tar entry format, the atime and ctime fields (unique to this format) need to be written as null characters if their value is the Unix Epoch. The bug was that we were writing them as "0" characters as other GNU metadata fields. This was causing tools like 7-zip to show an extra unexpected folder with a name formed by just "0" characters. Other tools were unable to read these archives as they were considered corrupt.
Added sync and async tests to validate directly in the memory stream that when atime and ctime represent the Unix Epoch, they are indeed written as null characters.

carlossanlop commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:
```diff
@@ -629,7 +629,7 @@ private int WriteCommonFields(Span<byte> buffer, TarEntryType actualEntryType)
             }
 
-            if (_dataStream != null && _size >= 0)
+            if (_size >= 0)
```
Yes. if the `_dataStream` is `null` we already know that `_size` is 0, and internally the `FormatNumeric` method won't try to access the `_data` stream.

ericstj commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:
```diff
@@ -1132,6 +1132,11 @@ private static int FormatOctal(long value, Span<byte> destination)
         {
+            if (timestamp == DateTimeOffset.UnixEpoch)
+            {
+                return 0;
```
This seems to write nothing at all and return 0 - is that what you meant by "nulls"?  Should it instead fill with 0 or can you count on the backing span being pre-filled with zeros?

carlossanlop replied:
Let me check. I'd be surprised if the byte buffer where we write everything is unexpectedly filled with garbage instead of zeroed by default.

carlossanlop replied:
In the async code we clear the rented buffer to zero it out, because they're not guaranteed to be clean:
https://github.com/dotnet/runtime/blob/518efe04f4f42ff1ec7d3a5a51b2a584ef58fa1d/src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarWriter.cs#L324

Looks like we do the same in the sync code although I don't recall if it's actually necessary (hence the lack of comment): https://github.com/dotnet/runtime/blob/518efe04f4f42ff1ec7d3a5a51b2a584ef58fa1d/src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarWriter.cs#L282

carlossanlop replied:
In the specific case of the ctime and atime fields, we don't want them to be filled with "0" chars. Instead we want them filled with nulls.

It's the size field the one that we want filled with "0" chars.

carlossanlop replied:
I believe I answered the concern, but please unresolve the conversation if you have more comments.

mmitche commented:
#1 and #2 fix in your PR description appear to be the same. Can you update?

carlossanlop commented:
> # 1 and # 2 fix in your PR description appear to be the same. Can you update?

Ah sorry. Fixed. The correct descriptions are in the commits themselves, anyway.

carlossanlop commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:
```diff
@@ -1132,6 +1132,11 @@ private static int FormatOctal(long value, Span<byte> destination)
         // Writes the specified DateTimeOffset's Unix time seconds, and returns its checksum.
         private int WriteAsTimestamp(DateTimeOffset timestamp, Span<byte> destination)
         {
+            if (timestamp == DateTimeOffset.UnixEpoch)
```
I thought this was obvious enough.

I should also add unit tests for the epoch being 0 and making sure the returned checksum matches the expectation.

carlossanlop replied:
I added a debug check. I also forked the method so that the timestamp check is only done when in Gnu.

carlossanlop commented:
Need to adjust the checksum unit tests, as the expected values are wrong now.

carlossanlop committed:
Create a separate timestamp writing method just for GNU timestamps. Regular timestamps (mtime) should not be affected by this.
Modify all the sync and async checksum shared code so that they reflect the changes in the expected values of size, atime and ctime.

carlossanlop commented:
I pushed an extra commit to separate the new timestamp checking logic from the old method and created another one just for the gnu timestamps. I also fixed all the checksum unit tests, sync and async, since the expected values for gnu were now different. I also separated the gnu checksum verification tests into their own separate ones since they need to additionaly verify the epoch values (no other formats will have this issue).
